ScreenCapture.NET/DirectX/DX11ScreenCapture.cs

```diff
@@ -11,370 +11,419 @@
+
+        ValidateCaptureZoneAndThrow(newX, newY, newUnscaledWidth, newUnscaledHeight);
 
-        private void InitializeCaptureZone(in CaptureZone captureZone)
+        captureZone.X = newX;
+        captureZone.Y = newY;
+
+        //TODO DarthAffe 01.05.2022: For now just reinitialize the zone in that case, but this could be optimized to only recreate the textures needed.
+        if ((width != null) || (height != null) || (downscaleLevel != null))
         {
-            Texture2DDescription stagingTextureDesc = new()
-            {
-                CpuAccessFlags = CpuAccessFlags.Read,
-                BindFlags = BindFlags.None,
-                Format = Format.B8G8R8A8_UNorm,
-                Width = captureZone.Width,
-                Height = captureZone.Height,
-                OptionFlags = ResourceOptionFlags.None,
-                MipLevels = 1,
-                ArraySize = 1,
-                SampleDescription = { Count = 1, Quality = 0 },
-                Usage = ResourceUsage.Staging
-            };
-            ID3D11Texture2D stagingTexture = _device!.CreateTexture2D(stagingTextureDesc);
-
-            ID3D11Texture2D? scalingTexture = null;
-            ID3D11ShaderResourceView? scalingTextureView = null;
-            if (captureZone.DownscaleLevel > 0)
+            (int newWidth, int newHeight) = CalculateScaledSize(newUnscaledWidth, newUnscaledHeight, newDownscaleLevel);
+            lock (_captureZones)
             {
-                Texture2DDescription scalingTextureDesc = new()
-                {
-                    CpuAccessFlags = CpuAccessFlags.None,
-                    BindFlags = BindFlags.RenderTarget | BindFlags.ShaderResource,
-                    Format = Format.B8G8R8A8_UNorm,
-                    Width = captureZone.UnscaledWidth,
-                    Height = captureZone.UnscaledHeight,
-                    OptionFlags = ResourceOptionFlags.GenerateMips,
-                    MipLevels = captureZone.DownscaleLevel + 1,
-                    ArraySize = 1,
-                    SampleDescription = { Count = 1, Quality = 0 },
-                    Usage = ResourceUsage.Default
-                };
-                scalingTexture = _device!.CreateTexture2D(scalingTextureDesc);
-                scalingTextureView = _device.CreateShaderResourceView(scalingTexture);
-            }
+                UnregisterCaptureZone(captureZone);
+
+                captureZone.UnscaledWidth = newUnscaledWidth;
+                captureZone.UnscaledHeight = newUnscaledHeight;
+                captureZone.Width = newWidth;
+                captureZone.Height = newHeight;
+                captureZone.DownscaleLevel = newDownscaleLevel;
+                captureZone.Buffer = new byte[newWidth * newHeight * BPP];
 
-            _captureZones[captureZone] = (stagingTexture, scalingTexture, scalingTextureView);
+                InitializeCaptureZone(captureZone);
```

Comment:
A bit similar to code on the RegisterCaptureZone method, maybe extracting to a common method? Unless you already have some thoughts on how to refactor this whole sector.

Reply:
There isn't really much in common here since it's the only time the values are set by assignment (this wasn't possible until now) and it requires a bigger rework anyway to get the possible performance benefits.